sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/RestrictionTracker.java

```diff
@@ -99,6 +100,10 @@
    */
   public abstract void checkDone() throws IllegalStateException;
 
+  public boolean isBounded() throws NotImplementedException {
+    throw new NotImplementedException("isBounded is not implemented.");
```

Comment:
If we want to make this optional, we should follow the `HasProgress` style and use an interface like `HasBoundedness`

We could also make this a required method and update the changelog stating that this is a breaking change. This would remove the "default" inference on the `@UnboundedPerElement`/`@BoundedPerElement` annotations on the DoFn.

Depending on the default ordering we want, we'll want to update update the restriction trackers in PeriodicSequence and Watch as well (note that Watch will have to be updated regardless since it has a tracker that converts unbounded restrictions to bounded ones).

Reply:
As discussed offline, we'll make `RestrictionTracker#isBounded` a required method and we will not use `@UnboundedPerElement`/`@BoundedPerElement` as part of the default inference.